crates/oxc_transformer/src/typescript/enum.rs

```diff
@@ -533,23 +535,68 @@ impl<'a, 'ctx> IdentifierReferenceRename<'a, 'ctx> {
         previous_enum_members: PrevMembers<'a>,
         ctx: &'ctx TraverseCtx<'a>,
     ) -> Self {
-        IdentifierReferenceRename { enum_name, enum_scope_id, ctx, previous_enum_members }
+        IdentifierReferenceRename {
+            enum_name,
+            previous_enum_members,
+            scope_stack: NonEmptyStack::new(enum_scope_id),
+            ctx,
+        }
     }
 }
 
 impl IdentifierReferenceRename<'_, '_> {
     fn should_reference_enum_member(&self, ident: &IdentifierReference<'_>) -> bool {
+        // Don't need to rename the identifier if it's not a member of the enum,
+        if !self.previous_enum_members.contains_key(&ident.name) {
+            return false;
+        };
+
         let symbol_table = self.ctx.scoping.symbols();
         let Some(symbol_id) = symbol_table.get_reference(ident.reference_id()).symbol_id() else {
-            // No symbol found. If the name is found in previous_enum_members,
-            // it must be referencing a member declared in a previous enum block: `enum Foo { A }; enum Foo { B = A }`
-            return self.previous_enum_members.contains_key(&ident.name);
+            // No symbol found, yet the name is found in previous_enum_members.
+            // It must be referencing a member declared in a previous enum block: `enum Foo { A }; enum Foo { B = A }`
+            return true;
         };
-        symbol_table.get_scope_id(symbol_id) == self.enum_scope_id
+
+        let symbol_scope_id = symbol_table.get_scope_id(symbol_id);
+        // Don't need to rename the identifier when it references a nested enum member:
+        //
+        // ```ts
+        // enum OuterEnum {
+        //   A = 0,
+        //   B = () => {
+        //     enum InnerEnum {
+        //       A = 0,
+        //       B = A,
+        //           ^ This references to `InnerEnum.A` should not be renamed
+        //     }
+        //     return InnerEnum.B;
+        //   }
+        // }
+        // ```
+        *self.scope_stack.first() == symbol_scope_id
+            // The resolved symbol is declared outside the enum,
+            // and we have checked that the name exists in previous_enum_members:
+            //
+            // ```ts
+            // const A = 0;
+            // enum Foo { A }
+            // enum Foo { B = A }
+            //                ^ This should be renamed to Foo.A
+            // ```
+        || !self.scope_stack.contains(&symbol_scope_id)
```

Comment:
@Dunqing  Maybe a wild idea: is it ok to assume that a inner scope id is always greater than a outer scope id (because the inner scope is created later)? If so we can just do `enum_scope_id >= symbol_scope_id` here.

Reply:
Actually, I think it will work, but it's not a good idea, because we can't guarantee that we won't break that assumption in the future.